akka-cluster-sharding-typed/src/main/scala/akka/cluster/sharding/typed/internal/ShardedDaemonProcessImpl.scala

```diff
@@ -42,37 +47,64 @@ private[akka] object ShardedDaemonProcessImpl {
   object KeepAlivePinger {
     sealed trait Event
     private case object Tick extends Event
-    private case object StartTick extends Event
+    private case object StartAllDone extends Event
 
     def apply[T](
         settings: ShardedDaemonProcessSettings,
         name: String,
         identities: Set[EntityId],
-        shardingRef: ActorRef[ShardingEnvelope[T]]): Behavior[Event] =
-      Behaviors.setup { context =>
-        Cluster(context.system).subscriptions ! Subscribe(
-          context.messageAdapter[SelfUp](_ => StartTick),
-          classOf[SelfUp])
-        Behaviors.withTimers { timers =>
-          def triggerStartAll(): Unit = {
-            identities.foreach(id => shardingRef ! StartEntity(id))
+        shardingRef: ActorRef[ShardingEnvelope[T]]): Behavior[Event] = {
+      val sortedIdentities = identities.toVector.sorted
+
+      def sendKeepAliveMessages()(implicit sys: ActorSystem[_]): Future[Done] = {
+        if (settings.keepAliveThrottleInterval == Duration.Zero) {
+          sortedIdentities.foreach(id => shardingRef ! StartEntity(id))
+          Future.successful(Done)
+        } else {
+          Source(sortedIdentities).throttle(1, settings.keepAliveThrottleInterval).runForeach { id =>
+            shardingRef ! StartEntity(id)
+          }
+        }
+      }
+
+      Behaviors.setup[Event] { context =>
+        implicit val system: ActorSystem[_] = context.system
+        val cluster = Cluster(system)
+
+        cluster.subscriptions ! Subscribe(context.messageAdapter[SelfUp](_ => Tick), classOf[SelfUp])
```

Comment:
You mean as an optimization to avoid the subscribe? It was like this before also and it will invoke the callback also when it's up. But, sure, I'll add the pre-check.

Reply:
I was thinking more that it could miss it, but maybe I just don't remember that we guarantee that you see a SelfUp even if already up when subscribing.